Ignore entropy from RAND_add()/RAND_seed() in FIPS mode

mspncp opened the pull request:
The functions RAND_add() and RAND_seed() provide a legacy API which
enables the application to seed the CSPRNG.

But NIST SP-800-90A clearly mandates that entropy *shall not* be provided
by the consuming application, neither for instantiation, nor for reseeding.

The provided random data will be mixed into the DRBG state as additional
data only, and no entropy will accounted for it.


##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->
- [ ] documentation is added or updated
- [ ] tests are added or updated

mspncp commented:
###  NIST-SP-800-90Ar1

#### 9.1 Instantiating a DRBG
...

**Required information not provided by the consuming application during instantiation**
(This information **shall not** be provided by the consuming application as an input parameter
during the instantiate request):

1. _entropy_input_: Input bits containing entropy. ...
2. _nonce_: A nonce as specified in Section 8.6.7. ...

(NIST SP-800-90Ar1, page 27)

#### 9.2 Reseeding a DRBG Instantiation
... 

**Required information not provided by the consuming application during reseeding** (This
information **shall not** be provided by the consuming application as an input parameter during the
reseed request):

1.	 entropy_input: Input bits containing entropy. ...
2.	 Internal state values ...

(NIST SP-800-90Ar1, page 30)

mspncp commented:
Marked as WIP, because documentation and tests are still missing.

Two questions from a lateral entrant:

- The `fips` configure option is still defunct  and reports "FIPS mode not supported" (since 1.1.0). Is there already a replacement, or does one currently still need to define `FIPS_MODE` manually for developing and testing?

- Is there already an agreement about how deviations from standard behaviour in FIPS mode should be documented in the manual pages? Will there be special FIPS sections in the manual pages, or shall the deviating behaviour be documented immediately following the description of the normal behaviour, marked with phrases like "In FIPS mode, _...deviating behaviour..._" ?

paulidale commented:
I think it tempting to always ignore the entropy in these calls irrespective of FIPS.  I've seen far too many cases where the supplied _entropy_ was anything but.  Engines can and providers will be able to supply seed material, they should cover hardware sources well enough.

As for the questions.  The intention is to not need a `fips` configure option.  The FIPS provider will be built in parallel.  So yes, a configure with `-DFIPS_MODE` is the way now.  Tests that use this macro will need to be updated to not and instead determine the mode at run time.  Early stages mean some things aren't ready yet.

I don't think we've discussed FIPS related documentation changes.

slontis commented:
The fips provider is bare bones at the moment, so RAND_XXX wont get pulled into that provider yet.
If you try adding -D FIPS_MODE it will fail to build currently.
https://github.com/openssl/openssl/pull/8662 attempts to add some hacky stubs, so that it still builds..

mattcaswell commented:
> The fips configure option is still defunct and reports "FIPS mode not supported" (since 1.1.0). Is there already a replacement, or does one currently still need to define FIPS_MODE manually for developing and testing?

Just to further expand on the comments above: the idea is that certain files will be compiled *twice* - once for inclusion in libcrypto, and once for inclusion in the FIPS module. When being compiled for inclusion in the FIPS module then "FIPS_MODE" will automatically be defined. We haven't yet pulled the RAND code into the FIPS module so you won't get the define yet without explicitly setting it manually.

paulidale commented:
If you are looking for a small project, attempting to define the RAND and entropy APIs would be a big help.

mspncp commented:
> I think it tempting to always ignore the entropy in these calls irrespective of FIPS. I've seen far too many cases where the supplied _entropy_ was anything but. Engines can and providers will be able to supply seed material, they should cover hardware sources well enough.


This reminds me of discussions we had way back in 2017 when I made my first large DRBG pull request(*). My initial approach was to use RAND_add() only for additional data, but it turned out that this would break too many existing applications. For years, RAND_add()/RAND_seed() had been the only way to explicitly add entropy without patching the source, and a lot of applications were using it (not only snake-oil vendors).

The best solution IMHO would be to deprecate RAND_add()/RAND_seed() starting with OpenSSL 3.0 and remove it in OpenSSL >= 4.0.


(*) Here a stroboscopic snapshot of the discussion I'm referring to:

- https://github.com/openssl/openssl/pull/4328#issuecomment-327614720
- https://github.com/openssl/openssl/pull/4328#issuecomment-327936005
- https://github.com/openssl/openssl/pull/4328#issuecomment-328163686
- https://github.com/openssl/openssl/pull/4328#issuecomment-328175416

mspncp commented:
> > The fips configure option is still defunct and reports "FIPS mode not supported" (since 1.1.0). Is there already a replacement, or does one currently still need to define FIPS_MODE manually for developing and testing?
> 
> Just to further expand on the comments above: the idea is that certain files will be compiled _twice_ - once for inclusion in libcrypto, and once for inclusion in the FIPS module. When being compiled for inclusion in the FIPS module then "FIPS_MODE" will automatically be defined. We haven't yet pulled the RAND code into the FIPS module so you won't get the define yet without explicitly setting it manually.

Then I would like to suggest to resurrect the `enable-fips/no-fips` configure option with the new semantics that it can be used by (e.g. non-US) developers to disable the compilation of the FIPS part, if they are not interested or don't need FIPS certified software. This would speed up the compilation process significantly.

It's ok for me if FIPS support is enabled by default, but it should be disablable.

mspncp commented:
@mattcaswell just for curiosity: how do you intend to handle the "code freeze" problem while at the same time being able to build FIPS and non-FIPS providers side-by-side? 

_(Hopefully not by sending DVDs via snail-mail? ;-) )_

mspncp commented:
> If you are looking for a small project, attempting to define the RAND and entropy APIs would be a big help.

Assuming you were a addressing me: Sounds like an interesting task, I can give it a thought...

paulidale commented:
> Assuming you were a addressing me: Sounds like an interesting task, I can give it a thought...

@mspncp I was.  I should have been more explicit.

mspncp commented:
Thanks @paulidale for your approval. Note however, that this pr is still WIP. The reason for that is the fact that some documentation is missing (the fact that application randomness is not accounted for in FIPS should be mentioned somewhere in the manual pages), and that IIRC there is a test reseeding via RAND_seed(), which will inevitably fail in FIPS mode and needs to be fixed. 

It has been lying around for a while, because I was not able to finish it before my vacation, and this week was very busy at work. So I remove the 'ready' label, but take your approval as polite reminder and encouragement to finish this pull request as soon as possible ;-).

mspncp commented:
Rebased and added documentation. Also enabled FIPS_MODE temporarily [to be removed] to demonstrate that the reseeding using RAND_add() fails in FIPS mode.

(Test still needs to be reworked.)

mspncp commented:
Squashed the fixups [without changing anything](https://github.com/openssl/openssl/compare/acd5975185c9d01f0f87d33682af192f69381770..a67163ff86f6b412709e0b3a57d98375a02807b3).

mspncp commented:
Commit 0c71c5907d11b32ef2d5df1d10adf739699e3941 fixes the drbgtest in FIPS mode. I'll wait until the CI's complete, then I remove the FIPS mode hack a67163ff86f6b412709e0b3a57d98375a02807b3 again and rerun the CI's without FIPS mode.

mspncp commented:
The testsuite succeeded in FIPS mode:

- AppVeyor: [master.24525](https://ci.appveyor.com/project/openssl/openssl/builds/24426845)
- Travis: [Build #25111](https://travis-ci.org/openssl/openssl/builds/530331106?utm_source=github_status&utm_medium=notification)

Reverting the FIPS hack and rerunning the CI's without FIPS mode. Ready now final review, provided the CI's succeed. All commits will be squashed.

mspncp commented:
> The testsuite succeeded in FIPS mode:
> 
> * AppVeyor: [master.24525](https://ci.appveyor.com/project/openssl/openssl/builds/24426845)
> * Travis: [Build #25111](https://travis-ci.org/openssl/openssl/builds/530331106?utm_source=github_status&utm_medium=notification)


And here the results without FIPS mode:

* AppVeyor: [master.24526](https://ci.appveyor.com/project/openssl/openssl/builds/24428432)
* Travis: [Build #251121](https://travis-ci.org/openssl/openssl/builds/530356510?utm_source=github_status&utm_medium=notification)

paulidale commented on crypto/rand/drbg_lib.c:
```diff
@@ -1101,8 +1101,20 @@ static int drbg_add(const void *buf, int num, double randomness)
+#endif
     if (buflen < seedlen || randomness < (double) seedlen) {
-#if defined(OPENSSL_RAND_SEED_NONE)
+# if defined(OPENSSL_RAND_SEED_NONE)
```
This shouldn't have the extra indent.
That or the #ifdef above should.

mspncp replied:
I reverted the reindentation.

paulidale commented on crypto/rand/drbg_lib.c:
```diff
@@ -1101,8 +1101,20 @@ static int drbg_add(const void *buf, int num, double randomness)
+    /*
+     * NIST SP-800-90A mandates that entropy *shall not* be provided
+     * by the consuming application, neither for instantiation, nor
+     * for reseeding. By setting the randomness to zero, we ensured
```
s/ensured/ensure/

mspncp replied:
done

paulidale commented on doc/man7/RAND_DRBG.pod:
```diff
@@ -264,6 +264,14 @@ from the trusted entropy sources.
 
+NOTE: Manual reseeding is *not allowed* in FIPS mode, because
+NIST SP-800-90A mandates that entropy *shall not* be provided by the
+consuming application, neither for instantiation, nor for reseeding.
```
Something seems funny with the two commas.  How about:
```
consuming application: neither for instantiation nor for reseeding.
```
instead?

mspncp replied:
I simply terminated the sentence after "application" and removed the rest, since it was dispensable.

paulidale replied:
I'm not seeing this change yet, but that's a good fix.

paulidale commented on doc/man7/RAND_DRBG.pod:
```diff
@@ -264,6 +264,14 @@ from the trusted entropy sources.
+consuming application, neither for instantiation, nor for reseeding.
+[NIST SP 800-90Ar1, Sections 9.1 and 9.2]. For that reason the B<randomness>
+argument is ignored and the random bytes provided by the L<RAND_add(3)> and
+L<RAND_seed(3)> calls are treated as additional data. (In other words, the
```
I don't think the parenthesized sentence is good.  Either remove the parentheses or remove the whole thing (I don't think it adds much to the paragraph).

mspncp replied:
I removed the sentence.

mspncp commented:
The change you were looking for is here: https://github.com/openssl/openssl/pull/8722/commits/d9c9f7b60a05fc68bad7bc391a488a226775f63d#diff-9181ac017a6177a5f2619f65c9b7a346R1107

paulidale commented:
Apologies, I missed it when I looked.  All good to go.

mspncp commented:
> Apologies, I missed it when I looked. All good to go.

Never mind, no problem.

_(FWIW: It might have been partly GitHub's fault, it has been a bit buggy the last few days: When I marked your annotations as 'resolved', I noticed that the one containing that specific change hadn't been marked 'outdated' and also didn't show the deletion. (It still doesn't, just unfold it.) The deletion only showed up in the 'Files changed' view.)_

paulidale commented:
It doesn't show as resolved here either :(

mspncp commented:
Merged to 3a50a8a91ad10e406544d16247957a17a5d5beac. Thank you for reviewing, Pauli!

mspncp commented:
> Merged to 3a50a8a.

Oh, no! I just noticed that the final commit which I pushed does not contain the deletion. This is really weird!  :-/

I need a little time to figure out what went wrong (i.e., whether it's a [layer 7 or layer 8](https://en.wikipedia.org/wiki/Layer_8) issue...), but I'm short on time now. I'll figure it out and fix it in a separate pull request...

paulidale commented:
That sounds okay.  I don't think it is critical.

mspncp commented:
> That sounds okay. I don't think it is critical.

The difference is not critical. But the fact, that my workflow failed, due to some (potential) GitHub issue, which I overlooked.
